inc/zoo/swar/SWAR.h

```diff
@@ -472,13 +481,53 @@ booleans(SWAR<NB, T> arg) noexcept {
 template<int NBits, typename T>
 constexpr auto
 differents(SWAR<NBits, T> a1, SWAR<NBits, T> a2) {
-    return booleans(a1 ^ a2);
+    return ~equals(a1, a2);
 }
 
+/**
+ * @return BooleanSWAR that contains a true value in each lane where the two
+ * inputs SWARs match.
+ */ 
+template<int NBits, typename T>
+constexpr BooleanSWAR<NBits, T>
+equals(SWAR<NBits, T> a1, SWAR<NBits, T> a2) noexcept {
+    // Knuth, TAOCP 4A pg 152
+    using S = swar::SWAR<NBits, T>;
+    const S x = (a1 ^ a2);
+    return BooleanSWAR<NBits, T>{S{S::MostSignificantBit} & ~(x| ((x|S{S::MostSignificantBit}) - S{S::LeastSignificantBit}))};
+}
+
+/**
+ * Less operations than equals()
+ * @return psuedoBooleanSWAR. If there are no null lanes present, will be all
+ * zeros.  If there is a null lane present, that lane will be all 1s, and any
+ * less significant lane will be all zeros. Anything more significant than the
+ * least significant null lane will be undefined.
+ * 
+ */
 template<int NBits, typename T>
 constexpr auto
-equals(SWAR<NBits, T> a1, SWAR<NBits, T> a2) {
-    return ~differents(a1, a2);
+firstZeroLane(SWAR<NBits, T> x) {
+    // Knuth, TAOCP 4A pg 152
+    using S = swar::SWAR<NBits, T>;
+    const auto h = S::MostSignificantBit, l = S::LeastSignificantBit;
+    return h & ( x.value() - l) & ~x.value();
```

Comment:
No!
Why doesn't this return a SWAR?
Why is this implementation not all SWAR?

Reply:
Returns auto, it'll SWAR as required, no?

Reply:
This returns `T`.
I don't know why you use `x.value()`

Reply:
```
const auto h = S{S::MostSignificantBit}, l = S{S::LeastSignificantBit};
```
We have to do that to ensure these computations are in SWAR type.  The constants we've got defined in SWAR I keep expecting to be SWAR themselves, but they never are, and it is annoying.  Did we have a reason for having them be T and not SWAR<NBits, T>?

Reply:
It is unclear which is better.
The constants, in the remote past, were used almost exclusively directly with `this->m_v`, then it was clear to keep them of the base type.  They are still implementation details.  Since they have proven to be very useful once you are working with proper SWAR types, I'm open to make them SWAR, and use `.value()` in the cases where they can't.

Reply:
I also strongly prefer keeping their type homogenous.
No need to change them now, I just hadn't realized I'd made the same irritating error over and over until today.
We can decide later.